### PR TITLE
[clang] static operators should evaluate object argument

### LINKED ISSUE #67976: Too aggresive C++23 static call operator optimization
Hello,

It looks like clang is too aggressive for C++23 static call optimization. Here is the simplified example of real code:

```cpp
#include <functional>
#include <iostream>

template <typename Tag>
struct Wrapper {
    template <typename F, typename... Args>
    static void operator()(F&& callable, Args&&... args) {
        Invoke(std::forward<F>(callable), std::forward<Args>(args)...);
    }

    template <typename F, typename... Args>
    static void Invoke(F&& callable, Args&&... args) {
        std::invoke(std::forward<F>(callable), std::forward<Args>(args)...);
    }
};

struct Manager {
    template <typename SingletonClass>
    static SingletonClass& Get() {
        std::cout << "get call\n";
        static SingletonClass instance;
        return instance;
    }
};

struct Tag {};

int main()
{
    using TaggedWrapper = Wrapper<Tag>;
    Manager::Get<TaggedWrapper>()([] { std::cout << "Function\n"; });
}
```

We assume that Manager::Get() will be called and `get call` will be printed, but it doesn't happen. If change call in main to `    Manager::Get<TaggedWrapper>().Invoke([] { std::cout << "Function\n"; });` it works as expected. `get cal` also will be printed if remove `static` from `operator()`. It reproducible even with `-O0`. clang-17.0.1

### PR DESCRIPTION
**Pull request by SuperSodaSea**
### Description

clang don't evaluate the object argument of `static operator()` and `static operator[]` currently, for example:

```cpp
#include <iostream>

struct Foo {
    static int operator()(int x, int y) {
        std::cout << "Foo::operator()" << std::endl;
        return x + y;
    }
    static int operator[](int x, int y) {
        std::cout << "Foo::operator[]" << std::endl;
        return x + y;
    }
};
Foo getFoo() {
    std::cout << "getFoo()" << std::endl;
    return {};
}
int main() {
    std::cout << getFoo()(1, 2) << std::endl;
    std::cout << getFoo()[1, 2] << std::endl;
}
```

`getFoo()` is expected to be called, but clang don't call it currently (17.0.2). This PR fixes this issue.

Fixes #67976.

### Walkthrough

- **clang/lib/Sema/SemaOverload.cpp**
  - **`Sema::CreateOverloadedArraySubscriptExpr` & `Sema::BuildCallToObjectOfClassType`**
  Previously clang generate `CallExpr` for static operators, ignoring the object argument. In this PR `CXXOperatorCallExpr` is generated for static operators instead, with the object argument as the first argument.
  - **`TryObjectArgumentInitialization`**
  `const` / `volatile` objects are allowed for static methods, so that we can call static operators on them.
- **clang/lib/CodeGen/CGExpr.cpp**
  - **`CodeGenFunction::EmitCall`**
  CodeGen changes for `CXXOperatorCallExpr` with static operators: emit and ignore the object argument first, then emit the operator call.
- **clang/lib/AST/ExprConstant.cpp**
  - **`‎ExprEvaluatorBase::handleCallExpr‎`**
  Evaluation of static operators in constexpr also need some small changes to work, so that the arguments won't be out of position.
- **clang/lib/Sema/SemaChecking.cpp**
  - **`Sema::CheckFunctionCall`**
  Code for argument checking also need to be modify, or it will fail the test `clang/test/SemaCXX/overloaded-operator-decl.cpp`.

### Tests

- **Added:**
    - **clang/test/AST/ast-dump-static-operators.cpp**
      Verify the AST generated for static operators.
    - **clang/test/SemaCXX/cxx2b-static-operator.cpp**
      Static operators should be able to be called on const / volatile objects.
- **Modified:**
    - **clang/test/CodeGenCXX/cxx2b-static-call-operator.cpp**
    - **clang/test/CodeGenCXX/cxx2b-static-subscript-operator.cpp**
      Matching the new CodeGen.

### Documentation

- **clang/docs/ReleaseNotes.rst**
  Update release notes.

### CODE REVIEW

**Comment by SuperSodaSea** [comment]
CC @dtcxzyw, @zygoloid

File: clang/lib/Sema/SemaOverload.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaOverload.cpp:14938]:
```suggestion
        // Initialize the object parameter.
```

File: clang/lib/AST/ExprConstant.cpp
Comment (shafik) [inline clang/lib/AST/ExprConstant.cpp:7810]: So in the static case, what value do we expect `ThisVal` to have and does that have consequences that the code after this is prepared for?

File: clang/lib/CodeGen/CGExpr.cpp
Comment (shafik) [inline clang/lib/CodeGen/CGExpr.cpp:5555]: nit
```suggestion
                 /*ParamsToSkip=*/0, Order);
```

File: clang/lib/CodeGen/CGExpr.cpp
Comment (shafik) [inline clang/lib/CodeGen/CGExpr.cpp:5558]: nit
```suggestion
                 E->getDirectCallee(), /*ParamsToSkip=*/0, Order);
```

File: clang/lib/AST/ExprConstant.cpp
Comment (SuperSodaSea) [reply clang/lib/AST/ExprConstant.cpp:7810]: It doesn't look like anything has gone wrong so far, but just to be on the safe side, maybe we could change the code below to this?

```diff
- This = &ThisVal;
+ if (MD->isInstance())
+   This = &ThisVal;
```

**Comment by SuperSodaSea** [comment]
Ping @cor3ntin @shafik

**Comment by dtcxzyw** [comment]
Ping.

**Comment by dtcxzyw** [comment]
Ping.

**Review by cor3ntin — COMMENTED** [review]
I think this looks fine, once we have a release note

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by SuperSodaSea** [comment]
Wait a minute, it failes to compile this situation:

```c++
struct Foo {
    static int operator()(int a, int b) { return a + b; }
};

void f() {
    const Foo foo;
    foo(1, 2); // 'this' argument to member function 'operator()' has type 'const Foo', but function is not marked const
}
```

Let me try to fix this...

**Comment by SuperSodaSea** [comment]
Problem solved :)

**Comment by SuperSodaSea** [comment]
Ping.

File: clang/docs/ReleaseNotes.rst
Comment (cor3ntin) [inline clang/docs/ReleaseNotes.rst:865]:
```suggestion
- Fix incorrect code generation caused by the object argument of ``static operator()`` and ``static operator[]`` calls not being evaluated.
```

File: clang/lib/Sema/SemaOverload.cpp
Comment (cor3ntin) [inline clang/lib/Sema/SemaOverload.cpp:5684]: The best wording i can find is this one https://eel.is/c++draft/over.match#funcs.general-4.sentence-5

```suggestion
  C++23 [over.match.funcs.general] p4.2 For static member functions, the implicit object parameter is considered to match any object.
```

**Review by cor3ntin — COMMENTED** [review]
Just a few nits, but I think it looks good

**Review by cor3ntin — APPROVED** [review]
LGTM
@erichkeane you want to look at that?

**Comment by erichkeane** [comment]
> LGTM @erichkeane you want to look at that?

I'd looked at it a while back, and just did another scroll, happy as-is.

File: clang/lib/Sema/SemaOverload.cpp
Comment (SuperSodaSea) [reply clang/lib/Sema/SemaOverload.cpp]: I checked the old version of the standard and found that this wording exists in C++98. File updated.

File: clang/lib/Sema/SemaOverload.cpp
Comment (shafik) [inline clang/lib/Sema/SemaOverload.cpp:5689]: Does this change have a functional change wrt this change or is this just a correctness fix?

Comment (SuperSodaSea) [reply clang/lib/Sema/SemaOverload.cpp:5689]: The following code won't compile without this change:

```c++
struct Foo {
    static int operator()(int a, int b) { return a + b; }
};

void f() {
    const Foo foo;
    foo(1, 2); // 'this' argument to member function 'operator()' has type 'const Foo', but function is not marked const
}
```

This issue was founded in [libc++ test suite](https://buildkite.com/llvm-project/clang-ci/builds/9164#018cd53e-cd20-4251-8d04-ddf463079deb/6-17), which calls static `operator()` on a const object. Test for this situation is added in [`clang/test/SemaCXX/cxx2b-static-operator.cpp`](https://github.com/llvm/llvm-project/blob/fdfa350fc07aaa06e6d30402b0b265ba3591fa51/clang/test/SemaCXX/cxx2b-static-operator.cpp).

**Comment by SuperSodaSea** [comment]
Ping.

**Comment by cor3ntin** [comment]
@SuperSodaSea I can merge this for you if both @MitalAshok and @shafik are happy with it!

**Review by shafik — APPROVED** [review]

**Comment by SuperSodaSea** [comment]
Ping @MitalAshok, are you happy with this PR?

**Review by cor3ntin — APPROVED** [review]
Feel free to land this if he doesn't reply within the day.

@AaronBallman can we backport that? It's a pretty nasty bug

**Comment by SuperSodaSea** [comment]
Should be easy to backport using llvmbot once it is merged.

File: clang/lib/CodeGen/CGExpr.cpp
Comment (AaronBallman) [inline clang/lib/CodeGen/CGExpr.cpp:5853]:
```suggestion  
  if (auto *OCE = dyn_cast<CXXOperatorCallExpr>(E)) {
```

File: clang/lib/CodeGen/CGExpr.cpp
Comment (AaronBallman) [inline clang/lib/CodeGen/CGExpr.cpp:5887]:
```suggestion
  unsigned NumParamsToSkip = 0;
  if (StaticOperator) {
    // If we're calling a static operator, we need to emit the object argument
    // and ignore it.
    EmitIgnoredExpr(E->getArg(0));
    NumParamsToSkip = 1;
  }
  EmitCallArgs(Args, dyn_cast<FunctionProtoType>(FnType), E->arguments(),
               E->getDirectCallee(), NumParamsToSkip, Order);
```

**Review by AaronBallman — APPROVED** [review]
LGTM with some minor nits. In terms of backporting, it's a bit more complicated than I'd like for a backport, but I don't see anything concrete to suggest this is a bad candidate. I'll give it a +1 for backporting, but I'd love a second opinion from @rjmccall or @erichkeane (or others) just to be safe.

File: clang/lib/CodeGen/CGExpr.cpp
Comment (SuperSodaSea) [reply clang/lib/CodeGen/CGExpr.cpp:5886]: I tried to use `ParamsToSkip` before, and got assertion error for type mismatch in test **cxx2b-static-call-operator.cpp** & **cxx2b-static-subscript-operator.cpp**:

```
# | Assertion failed: (isGenericMethod || Ty->isVariablyModifiedType() || Ty.getNonReferenceType()->isObjCRetainableType() || getContext() .getCanonicalType(Ty.getNonReferenceType()) .getTypePtr() == getContext().getCanonicalType((*Arg)->getType()).getTypePtr()) && "type mismatch in call argument!", file .../llvm-project/clang/lib/CodeGen/CGCall.cpp, line 4470
```

https://github.com/llvm/llvm-project/blob/22be6a2c66fceead40cf7561806bdaf424cd3c71/clang/lib/CodeGen/CGCall.cpp#L4463-L4470

So I have to use `drop_begin` instead.

Comment (AaronBallman) [reply clang/lib/CodeGen/CGExpr.cpp]: I think that suggests there's still a problem; we should not have to manually drop the arguments when there's a parameter explicitly for that. I think what's happening is that there's a mismatch between static call operator prototypes and the checking logic in `EmitCallArgs`. CC @efriedma-quic @rjmccall

Comment (efriedma-quic) [reply clang/lib/CodeGen/CGExpr.cpp]: I don't really see an issue.  ParamsToSkip skips declared parameters (i.e. parts of the callee's signature).  Here, we're skipping arguments (subexpressions of the call), so ParamsToSkip isn't helpful.  The object in question isn't part of the callee's signature in the first place.

Comment (AaronBallman) [reply clang/lib/CodeGen/CGExpr.cpp]: Ah, thank you Eli! I missed that this was a params vs args issue.

**Comment by AaronBallman** [comment]
LGTM now, thank you for bearing with me! Do you need one of us to commit this on your behalf?

**Comment by SuperSodaSea** [comment]
Yeah, I'd be happy if anyone with write access could help. I'll create a backport issue after the commit.

**Comment by AaronBallman** [comment]
> Yeah, I'd be happy if anyone with write access could help. I'll create a backport issue after the commit.

I've landed the changes, thank you for the fix and the offer to help backport!

**Comment by nico** [comment]
This might've broken clangd tests: http://45.33.8.238/linux/129484/step_9.txt

Does that look related? If so, please take a look and revert for now if it takes a while to fix.

**Comment by AaronBallman** [comment]
Hmmm that does look related. I'll revert so we get back to green. CC @sam-mccall in case he has opinions on how we can ease tension between making fixes in Clang that impact clangd tests (clangd is a consumer of Clang and we usually expect consumers to react when their tests break due to conforming and correct changes in Clang). I mostly want to be mindful of the new contributor experience, as in this case.

**Comment by erichkeane** [comment]
I have absolutely no idea how to fix it as I don't understand the syntax of the clangd tests, but it SEEMS like it is just a source-location type thing that is wrong?

https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/unittests/InlayHintTests.cpp#L810

From my read of the test output, it is 'l2' on line 831 (which is line 20 of the test best i can tell?) where the 'suggestion' is chars 6-8 instead of 9-10.

I have no idea what any of that means however, so hopefully someone can stop by and help out @SuperSodaSea .

**Comment by SuperSodaSea** [comment]
It seems that clangd doesn't like the new usage for the static operator of the CXXOperatorCallExpr, so the hint is misplaced. I'm looking into **clang-tools-extra/clangd/InlayHints.cpp** now...

**Comment by AaronBallman** [comment]
CC @zyn0217 as the original author of that test case and @HighCommander4 as the original reviewer.

**Comment by SuperSodaSea** [comment]
https://github.com/SuperSodaSea/llvm-project/blob/1ceaae47b2b43fd8fa5512e20e0b32a7e8f4ab5b/clang-tools-extra/clangd/InlayHints.cpp#L660-L664

```diff
      if (const CXXMethodDecl *Method =
              dyn_cast_or_null<CXXMethodDecl>(Callee.Decl))
-       if (Method->isInstance() &&
-           (IsFunctor || Method->hasCXXExplicitFunctionObjectParameter()))
+       if (IsFunctor || (Method->isInstance() &&
+                         Method->hasCXXExplicitFunctionObjectParameter()))
          Args = Args.drop_front(1);
```

Here is the code need to be modified (`!Method->isInstance() && IsFunctor` also need to drop the first argument). Passed `check-clangd` on my own build.

Commit: dea08ecc0d9caf47f7e9bdac113844b8b2bfb68e

**Comment by SuperSodaSea** [comment]
By the way, maybe we should add a clangd label to this PR?

File: clang/lib/Sema/SemaOverload.cpp
Comment (zyn0217) [inline clang/lib/Sema/SemaOverload.cpp:15105]: While I haven’t had time to go through this PR, these lines appear to be the reason for the test failure. We had been relying on these different CallExprs to tell whether we should drop the explicit object parameter at the clangd site (IsFunctor). And this seemingly breaks that expectation as well as changes the AST?

Comment (SuperSodaSea) [reply clang/lib/Sema/SemaOverload.cpp:15105]: Yeah, with this change the CXXOperatorCall will always have an operator argument, even if the operator is static, this is intended in this PR.

Comment (SuperSodaSea) [reply clang/lib/Sema/SemaOverload.cpp:15105]: In short, for the following code
```cpp
// struct Foo { static int operator(int a, int b); };
// Foo foo;
foo(1, 2);
```
, the AST is changed from
```
CallExpr
|- operator()
|- a
\- b
```
to
```
CXXOperatorCallExpr
|- operator()
|- foo
|- a
\- b
```
And other situations are not affected.

Comment (zyn0217) [reply clang/lib/Sema/SemaOverload.cpp:15105]: Thanks for the explanation! I think this is reasonable and makes it much more intuitive for static operator call expressions. So, if I understand the change correctly, I think we can simplify the condition testing at the clangd side: we can avoid the check isInstance on expressions. That means, can we just reduce the condition to IsFunctor && hasCXXExplicitFunctionObjectParameter?

I’m on my phone now, so I couldn't validate my thoughts. Could you please help me for it? I’m willing to help you reland the patch if that works!

Comment (SuperSodaSea) [reply clang/lib/Sema/SemaOverload.cpp:15105]: You are right, `hasCXXExplicitFunctionObjectParameter()` should imply `isInstance()`, so `isInstance()` is no longer needed. Passed `check-clangd` on my own build.

```diff
      if (const CXXMethodDecl *Method =
              dyn_cast_or_null<CXXMethodDecl>(Callee.Decl))
-       if (Method->isInstance() &&
-           (IsFunctor || Method->hasCXXExplicitFunctionObjectParameter()))
+       if (IsFunctor || Method->hasCXXExplicitFunctionObjectParameter())
          Args = Args.drop_front(1);
```

Diff: 1ceaae47b2b43fd8fa5512e20e0b32a7e8f4ab5b...910ae406c88efd5cc97894dc1408606094bc0b82

Comment (zyn0217) [reply clang/lib/Sema/SemaOverload.cpp:15105]: Thanks. Given that the commit has been reverted, can you please submit a new relanding PR?

**Comment by SuperSodaSea** [comment]
Relanding at #80108.

**Pull request merged**
